### PR TITLE
[management] Add support to ECDSA public Keys

### PR DESCRIPTION
**Pull request by HarryKodden**
### Checklist
- [ ] Is it a bug fix
- [ ] Is a typo/documentation fix
- [X] Is a feature enhancement
- [X] It is a refactor
- [ ] Created tests that fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

**Comment by HarryKodden** [comment]
@mlsmaycon Hi Maycon, I made a new PR and this is having no issues. Is this OK to proceed with ?

**Comment by mlsmaycon** [comment]
Thanks, @HarryKodden. Yes, can you reference it here and close the other one?

**Comment by HarryKodden** [comment]
This PR replaces https://github.com/netbirdio/netbird/pull/2101 since that was having merging issues.
2101 will be closed.

File: management/server/jwtclaims/jwtValidator.go [hunk 1]
```diff
@@ -205,83 +219,88 @@ func getPemKeys(ctx context.Context, keysLocation string) (*Jwks, error) {
...
-	if err != nil {
-		return "", fmt.Errorf("unable to decode JWK exponent, error: %s", err)
-	}
+func getPublicKeyFromECDSA(jwk JSONWebKey) (publicKey *ecdsa.PublicKey, err error) {
 
-	publicKey := &rsa.PublicKey{
-		N: intModules,
-		E: exponent,
+	if jwk.X == "" || jwk.Y == "" || jwk.Crv == "" {
+		return nil, fmt.Errorf("ecdsa key incompleet")
```
Comment (bcmmbaga) [inline management/server/jwtclaims/jwtValidator.go]:
```suggestion
		return nil, fmt.Errorf("ecdsa key incomplete")
```

File: management/server/jwtclaims/jwtValidator.go [hunk 2]
```diff
@@ -115,15 +128,16 @@ func NewJWTValidator(ctx context.Context, issuer string, audienceList []string,
 				}
 			}
 
-			cert, err := getPemCert(ctx, token, keys)
+			publicKey, err := getPublicKey(token, keys)
 			if err != nil {
+				log.Errorf("getPublicKey error: %s", err)
```
Comment (bcmmbaga) [inline management/server/jwtclaims/jwtValidator.go]: Let's log the error with context

```suggestion
				log.WithContext(ctx).Errorf("cannot get PublicKey: %s", err)
```

File: management/server/jwtclaims/jwtValidator.go [hunk 3]
```diff
@@ -115,15 +128,16 @@ func NewJWTValidator(ctx context.Context, issuer string, audienceList []string,
...
 				return nil, err
 			}
 
-			result, _ := jwt.ParseRSAPublicKeyFromPEM([]byte(cert))
-			return result, nil
+			return publicKey, nil
 		},
-		SigningMethod:       jwt.SigningMethodRS256,
+		// Canged SigningMethod to 'nil', since we can't tell yet...
+		SigningMethod:       nil, // jwt.SigningMethodRS256,
```
Comment (bcmmbaga) [inline management/server/jwtclaims/jwtValidator.go]: We still needs to add the correct signing method here. As `nil` will invalidate this check https://github.com/netbirdio/netbird/blob/ddea0011709f091acd902d871822187f75fb091d/management/server/jwtclaims/jwtValidator.go#L165-L171

File: management/server/jwtclaims/jwtValidator.go [hunk 4]
```diff
@@ -205,83 +219,88 @@ func getPemKeys(ctx context.Context, keysLocation string) (*Jwks, error) {
...
+		if jwks.Keys[k].Kty == "EC" {
+			log.Debugf("generating PublicKey from ECDSA JWK")
+			return getPublicKeyFromECDSA(jwks.Keys[k])
+		}
 
-func generatePemFromJWK(jwk JSONWebKey) (string, error) {
-	decodedModulus, err := base64.RawURLEncoding.DecodeString(jwk.N)
-	if err != nil {
-		return "", fmt.Errorf("unable to decode JWK modulus, error: %s", err)
+		log.Debugf("Key Type: %s not yet supported, please raise ticket !", jwks.Keys[k].Kty)
```
Comment (bcmmbaga) [inline management/server/jwtclaims/jwtValidator.go]: Let's log the error with context

```suggestion
		log.WithContext(ctx).Debugf("Key Type: %s not yet supported, please raise ticket!", jwks.Keys[k].Kty)
```

File: management/server/jwtclaims/jwtValidator.go [hunk 5]
```diff
@@ -205,83 +219,88 @@ func getPemKeys(ctx context.Context, keysLocation string) (*Jwks, error) {
...
+			return jwt.ParseRSAPublicKeyFromPEM([]byte(cert))
 		}
-		log.WithContext(ctx).Debugf("generating validation pem from JWK")
-		return generatePemFromJWK(jwks.Keys[k])
-	}
 
-	return cert, errors.New("unable to find appropriate key")
-}
+		if jwks.Keys[k].Kty == "RSA" {
+			log.Debugf("generating PublicKey from RSA JWK")
```
Comment (bcmmbaga) [inline management/server/jwtclaims/jwtValidator.go]: Same here. Lets add log with context

File: management/server/jwtclaims/jwtValidator.go [hunk 6]
```diff
@@ -205,83 +219,88 @@ func getPemKeys(ctx context.Context, keysLocation string) (*Jwks, error) {
...
+			return getPublicKeyFromRSA(jwks.Keys[k])
+		}
+		if jwks.Keys[k].Kty == "EC" {
+			log.Debugf("generating PublicKey from ECDSA JWK")
```
Comment (bcmmbaga) [inline management/server/jwtclaims/jwtValidator.go]: Same here. Lets add log with context

**Review by bcmmbaga — CHANGES_REQUESTED** [review]

File: management/server/jwtclaims/jwtValidator.go (hunk 3 above)
Comment (HarryKodden) [reply management/server/jwtclaims/jwtValidator.go]: That is true, But the point is, You can't expect that the token is signed with a specific algorithm unless the OIDC Client is registered specifically with the instruction to use that specific algorithms for token signing. If that was not explicitly instructed during client registration, then the OIDC Provider just picks one algorithm and you can see in the header the algorithm which one was used. My 2-cents: The code is unnecessary strict to enforce a specific algorithm.

Comment (bcmmbaga) [reply management/server/jwtclaims/jwtValidator.go]: we're not hardcoding the algorithm but rather extracting it dynamically from the JWK that matches the token.

For instance, we can update the `getPublicKey`  to return both the public key and the signing method:
```go
func getPublicKey(token *jwt.Token, jwks *Jwks) (interface{}, jwt.SigningMethod, error) {}
```
This allows us to validate the token using the correct signing method in `ValidationKeyGetter`, ensuring that the process is flexible and aligns with the algorithm specified in the token header.

Comment (HarryKodden) [reply management/server/jwtclaims/jwtValidator.go]: @bcmmbaga I am afraid I do not see your point. We could off course return the SigingMethod to look at the alg value in the jwt token. But then the ValidateKeyGetter is comparing the same 'alg' value (which is off course equal...)
Are we on the same page ?

Comment (bcmmbaga) [reply management/server/jwtclaims/jwtValidator.go]: I understand your point now. In that case, let's go ahead and remove the `SigningMethod` as well as the check for the algo. Then we can merge this.

Comment (HarryKodden) [reply management/server/jwtclaims/jwtValidator.go]: Agreed, ~~Will do !~~ Done !

**Pull request merged**